### PR TITLE
CAD-753:  refactors & detecting lost/martian Txs

### PR DESCRIPTION
**Pull request by deepfire**
1.  refactor & add missing/martian analysis
1.  cluster3nodes:  configure to log Tx receipts
1.  make the explorer DB cleanup non-invasive, so DB permissions don't need re-initialisation
1.  cluster3nodes:  fix tmux invocation
1.  cluster3nodes:  allow Nix usage
1.  cluster3nodes:  fix the need to remain-on-exit
1.  analyse.sh:  refactor & add lost tx counting
1.  misc stuff, whitespace cleanup, etc

### CODE REVIEW

File: benchmarks/cluster3nodes/run-3node-cluster.sh [hunk 1]
```diff
@@ -1,4 +1,7 @@
 #!/usr/bin/env bash
+# shellcheck disable=SC1090
+
+. "$(realpath "$(dirname "$0")")"/lib.sh
```
Comment (CodiePP) [inline benchmarks/cluster3nodes/run-3node-cluster.sh]: why not move $BASEPATH here?

File: benchmarks/cluster3nodes/run_explorer.sh [hunk 2]
```diff
@@ -1,4 +1,7 @@
 #!/bin/sh
+# shellcheck disable=SC1090
+
+. "$(realpath "$(dirname "$0")")"/lib.sh
 
 BASEDIR=$(realpath $(dirname $0))
```
Comment (CodiePP) [inline benchmarks/cluster3nodes/run_explorer.sh]: swap the position of the two lines?

File: scripts/analyse.sh [hunk 3]
```diff
@@ -1,37 +1,93 @@
...
+# explorer_log_prefix="${2:-node-on-explorer/node}"
+generator_log_prefix="${1:-logs/generator}"
+explorer_log_prefix="${2:-logs/node}"
 
-# extract txs from generator
-${BP}/xtx.sh generator/tx-gen-*.json > txs
-sort -k 1 -t ';' txs > txs1
+mkdir -p 'analysis'
+cd 'analysis'
+rm -f ./*
```
Comment (CodiePP) [inline scripts/analyse.sh:13]: this is kind of dangerous! imagine that the directory could not be created and cd into it. OK, we probably do not have access right then..

File: scripts/analyse.sh [hunk 4]
```diff
@@ -1,37 +1,93 @@
...
-  ${BP}/xbk.sh $L >> bks
-done
+sender_logs=($(ls ../"${generator_log_prefix}"-*.json))
+receiver_logs=($(ls ../"${explorer_log_prefix}"-*.json))
 
-sort -k 2 -t ';' bks > bks1
+###
+### Since we're using intermediate files with intricate semantics of sorting,
+### we better use more descriptive file names, let we go awry..
+###
```
Comment (CodiePP) [inline scripts/analyse.sh:22]: good idea! is certainly helpful

File: scripts/analyse.sh [hunk 5]
```diff
@@ -1,37 +1,93 @@
...
+-- Sends:                analysis/stx:stime.1
+-- Receipts:             analysis/rtx:rtime.1
+EOF
+if test "${count_sent}" != "${count_recvd}"
+then join -1 1       -2 1 -v 1 -t ";" \
+          stx:stime.1 rtx:rtime.1          > rtx:stime:rtime-missing.1
+     join -1 1       -2 1 -v 2 -t ";" \
+          stx:stime.1 rtx:rtime.1          > rtx:stime:rtime-martian.1
+     missing=$(cat rtx:stime:rtime-missing.1 | wc -l)
+     martian=$(cat rtx:stime:rtime-martian.1 | wc -l)
```
Comment (CodiePP) [inline scripts/analyse.sh:62]: why "martian"? looks like the same as "missing"

File: scripts/grep-mem.sh [hunk 6]
```diff
@@ -1,4 +1,3 @@
 #!/bin/sh
 
-jq -r 'select(.data.kind=="LogValue" and .data.name=="Mem.resident") | [ .at, .data.name, .data.value.contents*4096 ] | @csv' $* | sed -e 's/\([0-9-]\+\)T\([0-9:.]\+\)Z/\1 \2/' 
-
+jq -r 'select(.data.kind=="LogValue" and .data.name=="Mem.resident") | [ .at, .data.name, .data.value.contents*4096 ] | @csv' $* | sed -e 's/\([0-9-]\+\)T\([0-9:.]\+\)Z/\1 \2/'
```
Comment (CodiePP) [inline scripts/grep-mem.sh]: difference?

File: scripts/grep-cpu.sh [hunk 7]
```diff
@@ -1,4 +1,3 @@
 #!/bin/sh
 
-jq -r 'select(.data.kind=="LogValue" and .data.name=="Stat.utime") | [ .at, .data.name, .data.value.contents ] | @csv' $* | sed -e 's/\([0-9-]\+\)T\([0-9:.]\+\)Z/\1 \2/' 
-
+jq -r 'select(.data.kind=="LogValue" and .data.name=="Stat.utime") | [ .at, .data.name, .data.value.contents ] | @csv' $* | sed -e 's/\([0-9-]\+\)T\([0-9:.]\+\)Z/\1 \2/'
```
Comment (CodiePP) [inline scripts/grep-cpu.sh]: difference?

File: shell.nix [hunk 8]
```diff
@@ -30,17 +30,17 @@ let
     # These programs will be available inside the nix-shell.
     buildInputs = with haskellPackages; [
```
Comment (CodiePP) [inline shell.nix]: just sorted?

File: benchmarks/cluster3nodes/analysis/timetoblock.lst [hunk 9]
```diff
@@ -0,0 +1 @@
+stx:stime:rtime:blk.2
```
Comment (CodiePP) [inline benchmarks/cluster3nodes/analysis/timetoblock.lst]: this looks wrong

**Review by CodiePP — CHANGES_REQUESTED** [review]
a lot of open questions
this looks like an early stage WIP

File: scripts/analyse.sh (hunk 4 above)
Comment (CodiePP) [reply scripts/analyse.sh:22]: is the naming schema also valid on other platforms? I could sense that Windows will not be happy with a ";" in between. So a simple "_" would be safer

File: scripts/analyse.sh (hunk 3 above)
Comment (deepfire) [reply scripts/analyse.sh:13]: if `cd` doesn't work, then `-e` above will terminate the script, and the `rm` will not happen -- so I think it's safe.

File: shell.nix (hunk 8 above)
Comment (deepfire) [reply shell.nix]: Yes, just sorted.

File: benchmarks/cluster3nodes/analysis/timetoblock.lst (hunk 9 above)
Comment (deepfire) [reply benchmarks/cluster3nodes/analysis/timetoblock.lst]: Dropped.

File: benchmarks/cluster3nodes/run-3node-cluster.sh (hunk 1 above)
Comment (deepfire) [reply benchmarks/cluster3nodes/run-3node-cluster.sh]: Good point, done!

File: benchmarks/cluster3nodes/run_explorer.sh (hunk 2 above)
Comment (deepfire) [reply benchmarks/cluster3nodes/run_explorer.sh]: Done.

File: scripts/analyse.sh (hunk 4 above)
Comment (deepfire) [reply scripts/analyse.sh:22]: Good point about Windows, that certainly should break! Fixed.

File: scripts/grep-mem.sh (hunk 6 above)
Comment (deepfire) [reply scripts/grep-mem.sh]: Trailing whitespace.. : -)

File: scripts/grep-cpu.sh (hunk 7 above)
Comment (deepfire) [reply scripts/grep-cpu.sh]: Again, trailing whitespace..

**Review by CodiePP — APPROVED** [review]
:+1:

**Pull request merged**